streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java

```diff
@@ -102,4 +139,232 @@ static class RawAndDeserializedValue<ValueType> {
+    protected <R> QueryResult<R> runTimestampedKeyQuery(final Query<R> query,
+                                                      final PositionBound positionBound,
+                                                      final QueryConfig config) {
+        final QueryResult<R> result;
+        final TimestampedKeyQuery<K, V> typedKeyQuery = (TimestampedKeyQuery<K, V>) query;
+        final KeyQuery<Bytes, byte[]> rawKeyQuery =
+                KeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+        final QueryResult<byte[]> rawResult =
+                wrapped().query(rawKeyQuery, positionBound, config);
+        if (rawResult.isSuccess()) {
+            final Function<byte[], ValueAndTimestamp<V>> deserializer = getDeserializeValue(serdes, wrapped());
+            final ValueAndTimestamp<V> value = deserializer.apply(rawResult.getResult());
+            final QueryResult<ValueAndTimestamp<V>> typedQueryResult =
+                    InternalQueryResultUtil.copyAndSubstituteDeserializedResult(rawResult, value);
+            result = (QueryResult<R>) typedQueryResult;
+        } else {
+            // the generic type doesn't matter, since failed queries have no result set.
+            result = (QueryResult<R>) rawResult;
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    protected <R> QueryResult<R> runTimestampedRangeQuery(final Query<R> query,
+                                                        final PositionBound positionBound,
+                                                        final QueryConfig config) {
+
+        final QueryResult<R> result;
+        final TimestampedRangeQuery<K, V> typedQuery = (TimestampedRangeQuery<K, V>) query;
+        RangeQuery<Bytes, byte[]> rawRangeQuery;
+        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        if (typedQuery.lowerBound().isPresent() && typedQuery.upperBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withRange(
+                    keyBytes(typedQuery.lowerBound().get()),
+                    keyBytes(typedQuery.upperBound().get())
+            );
+        } else if (typedQuery.lowerBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withLowerBound(keyBytes(typedQuery.lowerBound().get()));
+        } else if (typedQuery.upperBound().isPresent()) {
+            rawRangeQuery = RangeQuery.withUpperBound(keyBytes(typedQuery.upperBound().get()));
+        } else {
+            rawRangeQuery = RangeQuery.withNoBounds();
+        }
+        if (!isKeyAscending) {
+            rawRangeQuery = rawRangeQuery.withDescendingKeys();
+        }
+        final QueryResult<KeyValueIterator<Bytes, byte[]>> rawResult =
+                wrapped().query(rawRangeQuery, positionBound, config);
+        if (rawResult.isSuccess()) {
+            final KeyValueIterator<Bytes, byte[]> iterator = rawResult.getResult();
+            final KeyValueIterator<K, V> resultIterator = new MeteredKeyValueTimestampedIterator(
```

Comment:
Return type should be ` KeyValueIterator<K, ValueAndTimestamp<V>>`, right?

Reply:
Yes, I will cast it to  `KeyValueIterator<K, ValueAndTimestamp<V>>`